vermouth/ismags.py

```diff
@@ -0,0 +1,1030 @@
+                    # Valid candidate
+                    candidates[sgn].add(gn)
+        return candidates
+
+    def largest_common_subgraph(self, symmetry=True):
+        """
+        Find the largest common induced subgraphs between :attr:`subgraph` and
+        :attr:`graph`.
+
+        Parameters
+        ----------
+        symmetry: bool
+            Whether symmetry should be taken into account. If False, found
+            largest common subgraphs may be symmetrically equivalent.
+
+        Yields
+        ------
+        dict
+            The found isomorphism mappings of {graph_node: subgraph_node}.
+        """
+        # The networkx VF2 algorithm is slightly funny in when it yields an
+        # empty dict and when not.
+        if not self.subgraph:
+            yield {}
+            return
+        elif not self.graph:
+            return
+
+        if symmetry:
+            _, cosets = self.analyze_symmetry(self.subgraph,
+                                              self._sgn_partitions,
+                                              self._sge_colors)
+            constraints = self._make_constraints(cosets)
+        else:
+            constraints = []
+
+        candidates = self._find_nodecolor_candidates()
+
+        if any(candidates.values()):
+            yield from self._largest_common_subgraph(candidates, constraints)
+        else:
+            return
+
+    def analyze_symmetry(self, graph, node_partitions, edge_colors):
+        """
+        Find a minimal set of permutations and corresponding co-sets that
+        describe the symmetry of :attr:`subgraph`.
+
+        Returns
+        -------
+        tuple[set[frozenset], dict]
+            The found permutations and co-sets. Permutations is a set of
+            frozenset of pairs of node keys which can be exchanged without
+            changing :attr:`subgraph`. The co-sets is a dictionary of node key:
+            set of node keys. Every key, value describes which values can be
+            interchanged while keeping all nodes less than the key the same.
+        """
+        key = hash((tuple(graph.nodes), tuple(graph.edges),
+                    tuple(map(tuple, node_partitions)), tuple(edge_colors.items())))
+        if key in self._symmetry_cache:
+            return self._symmetry_cache[key]
+        node_partitions = list(self._refine_node_partitions(graph,
+                                                            node_partitions,
+                                                            edge_colors))
+        assert len(node_partitions) == 1
+        node_partitions = node_partitions[0]
+        permutations, cosets = self._process_opp(graph,
+                                                 node_partitions,
+                                                 node_partitions,
+                                                 edge_colors)
+        self.__class__._symmetry_cache[key] = permutations, cosets

```

Comment:
Why is the cache at the class and not the instance level?

Reply:
Partly because of the API I inherit from NX. You need to make a new object for every graph/subgraph pair, which would wipe the cache after every run. It would be much better/saner/cleaner if we could make an ISMAGS object once, and ask it for the isomorphisms between two graphs.

Reply:
OK. I do not like it, but I kind of understand. The problem I see is that it can be quite a memory leak in the way it is implemented here. You are never going to clean your many permutations and co-sets because you will always have a reference to them in that cache.

Reply:
Good point, I hadn't considered that. It'll really depend on the graphs in question whether it'll be bad or not. At least I only store a hash of the graph instead of the graph itself as key.